javascript/extractor/src/com/semmle/js/dependencies/Fetcher.java

```diff
@@ -0,0 +1,147 @@
+        synchronized (mkdirpLock) {
+            Files.createDirectories(dir);
+        }
+    }
+
+    private static final Pattern validPackageName = Pattern.compile("(?:@[\\w.-]+\\/)?\\w[\\w.-]*");
+
+    private boolean isValidPackageName(String name) {
+        return validPackageName.matcher(name).matches();
+    }
+
+    public static Path toSafePath(String string) {
+        if (string == null) return null;
+        Path path = Paths.get(string).normalize();
+        if (path.startsWith("..") || path.isAbsolute()) {
+            return null;
+        }
+        return path;
+    }
+
+    /**
+     * Submits a GET request to the given URL and returns an input with the response.
+     */
+    private InputStream fetch(String url) throws IOException {
+        URLConnection connection = new URL(url).openConnection();
+        connection.setRequestProperty("Accept-Encoding", "gzip, identity, *");
+        connection.setDoInput(true);
+        connection.connect();
+        InputStream input = connection.getInputStream();
+        String encoding = connection.getContentEncoding();
+        if ("gzip".equals(encoding)) {
+            return new GzipCompressorInputStream(new BufferedInputStream(input));
+        } else {
+            return input;
+        }
+    }
+
+    /**
+     * Fetches the packument for the given package (containing all versions of the package.json).
+     */
+    public Packument getPackument(String packageName) throws IOException {
+        if (!isValidPackageName(packageName)) {
+            throw new IOException("Package name contains unexpected characters:" + packageName);
+        }
+        System.out.println("Fetching package metadata for " + packageName);
+        try (Reader reader = new BufferedReader(new InputStreamReader(fetch("https://registry.npmjs.org/" + packageName)))) {
+            return new Gson().fromJson(reader, Packument.class);
+        }
+    }
+
+    /**
+     * Extracts the package at the given tarball URL into the given directory.
+     * <p>
+     * Only `package.json` and `.d.ts` files are extracted.
+     *
+     * @return paths of the files created by this call, relative to <code>destDir</code>
+     */
+    public List<Path> extractFromTarballUrl(String tarballUrl, Path destDir) throws IOException {
+        if  (!tarballUrl.startsWith("https://registry.npmjs.org/") || !tarballUrl.endsWith(".tgz")) { // Paranoid check
+            throw new IOException("Tarball URL has unexpected format: " + tarballUrl);
+        }
+        System.out.println("Unpacking " + tarballUrl + " to " + destDir);
+        List<Path> relativePaths = new ArrayList<>();
+        try (InputStream rawStream = new URL(tarballUrl).openStream()) {
+            // Despite having the .tgz extension, the file is not always gzipped, sometimes it's just a raw tar archive,
+            // regardless of what Accept-Encoding header we send.
+            // Sniff the header to detect which is the case.
```

Comment:
I'll add a comment to clarify why, but unfortunately not. The two streams are compressed for very different reasons and require different means for determining the compression format.

The packument is gzipped because we use the `Accept-Encoding` header, i.e. it's compressed as part of the HTTP layer. The compression format can thus be deduced from the `Content-Encoding` header.

For the tarball, the compression format is not determined by the `Accept-Encoding` header. Setting `Accept-Encoding` here has no effect and the `Content-Encoding` is never `gzip`. Think of it as a file sent via an uncompressed HTTP stream, and that file happens to be gzip-compressed in some cases.